File: sentil-java/src/main/java/io/github/sedislab/sentil/RingBuffer.java
package io.github.sedislab.sentil;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.OptionalDouble;

/** A fixed-capacity ring buffer of timed samples with running statistics. */
public final class RingBuffer extends NativeResource {
    RingBuffer(long handle) {
        super(handle, NativeLib::ringBufferDestroy);
    }

    /** A buffer that holds at most capacity samples. */
    public static RingBuffer create(long capacity) throws SentilException {
        return new RingBuffer(NativeLib.ringBufferCreate(capacity));
    }

    /** Append a sample, returning the evicted oldest one on overflow. */
    public Optional<Sample> push(double time, double value) throws SentilException {
        Sample evicted = NativeLib.ringBufferPush(handle(), time, value);
        return evicted.found() ? Optional.of(evicted) : Optional.empty();
    }

    /** Drop every sample, keeping the capacity. */
    public void clear() {
        NativeLib.ringBufferClear(handle());
    }

    /** The number of samples currently held. */
    public long length() {
        return NativeLib.ringBufferLen(handle());
    }

    /** The most samples the buffer can hold. */
    public long capacity() {
        return NativeLib.ringBufferCapacity(handle());
    }

    public boolean isEmpty() {
        return NativeLib.ringBufferIsEmpty(handle());
    }

    public boolean isFull() {
        return NativeLib.ringBufferIsFull(handle());
    }

    /** The oldest sample. */
    public Sample front() {
        return NativeLib.ringBufferFront(handle());
    }

    /** The newest sample. */
    public Sample back() {
        return NativeLib.ringBufferBack(handle());
    }

    /** The sample at an index counted from the oldest. */
    public Sample get(long index) {
        return NativeLib.ringBufferGet(handle(), index);
    }

    /** Remove and return the oldest sample. */
    public Sample popFront() {
        return NativeLib.ringBufferPopFront(handle());
    }

    /** Remove and return the newest sample. */
    public Sample popBack() {
        return NativeLib.ringBufferPopBack(handle());
    }

    /** The held sample whose time is nearest the query. */
    public Sample closestToTime(double time) {
        return NativeLib.ringBufferClosestToTime(handle(), time);
    }

    private static OptionalDouble optional(double[] result) {
        return result.length == 0 ? OptionalDouble.empty() : OptionalDouble.of(result[0]);
    }

    /** The running mean of the buffered values. */
    public OptionalDouble mean() {
        return optional(NativeLib.ringBufferMean(handle()));
    }

    /** The running variance, empty until two samples are held. */
    public OptionalDouble variance() {
        return optional(NativeLib.ringBufferVariance(handle()));
    }

    /** The running standard deviation, empty until two samples are held. */
    public OptionalDouble stdDev() {
        return optional(NativeLib.ringBufferStdDev(handle()));
    }

    /** The smallest buffered value. */
    public OptionalDouble min() {
        return optional(NativeLib.ringBufferMin(handle()));
    }

    /** The largest buffered value. */
    public OptionalDouble max() {
        return optional(NativeLib.ringBufferMax(handle()));
    }

    /** Recompute the running mean and variance from scratch. */
    public void recomputeStatistics() {
        NativeLib.ringBufferRecomputeStatistics(handle());
    }

    /** The value recorded at the query time, within a small tolerance. */
    public OptionalDouble atTime(double time) {
        return optional(NativeLib.ringBufferAtTime(handle(), time));
    }

    /** The earliest and latest times held, as [start, end]. */
    public Optional<double[]> timeRange() {
        double[] range = NativeLib.ringBufferTimeRange(handle());
        return range.length == 0 ? Optional.empty() : Optional.of(range);
    }

    /** The samples whose time lies in [start, end], oldest first. */
    public List<Sample> between(double start, double end) throws SentilException {
        return Arrays.asList(NativeLib.ringBufferBetween(handle(), start, end));
    }
}

File: sentil-matlab/+sentil/RingBuffer.m
classdef RingBuffer < handle
    %RINGBUFFER A fixed-capacity rolling window of timed samples with running statistics.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
    end

    methods
        function obj = RingBuffer(capacity)
            obj.Handle = sentil_mex('ring_buffer_create', double(capacity));
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('ring_buffer_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
        end

        function evicted = push(obj, time, value)
            %PUSH Append a timed sample, returning any evicted sample.
            obj.assertOpen();
            evicted = sentil_mex('ring_buffer_push', obj.Handle, double(time), double(value));
        end

        function clear(obj)
            %CLEAR Drop every sample.
            obj.assertOpen();
            sentil_mex('ring_buffer_clear', obj.Handle);
        end

        function n = length(obj)
            %LENGTH The number of samples held.
            obj.assertOpen();
            n = sentil_mex('ring_buffer_len', obj.Handle);
        end

        function c = capacity(obj)
            %CAPACITY The maximum number of samples.
            obj.assertOpen();
            c = sentil_mex('ring_buffer_capacity', obj.Handle);
        end

        function tf = is_empty(obj)
            obj.assertOpen();
            tf = sentil_mex('ring_buffer_is_empty', obj.Handle);
        end

        function tf = is_full(obj)
            obj.assertOpen();
            tf = sentil_mex('ring_buffer_is_full', obj.Handle);
        end

        function s = front(obj)
            %FRONT The oldest sample, or [].
            obj.assertOpen();
            s = sentil_mex('ring_buffer_front', obj.Handle);
        end

        function s = back(obj)
            %BACK The newest sample, or [].
            obj.assertOpen();
            s = sentil_mex('ring_buffer_back', obj.Handle);
        end

        function s = get(obj, index)
            %GET The 1-based i-th sample from the front, or [].
            obj.assertOpen();
            s = sentil_mex('ring_buffer_get', obj.Handle, double(index));
        end

        function s = pop_front(obj)
            %POP_FRONT Remove and return the oldest sample, or [].
            obj.assertOpen();
            s = sentil_mex('ring_buffer_pop_front', obj.Handle);
        end

        function s = pop_back(obj)
            %POP_BACK Remove and return the newest sample, or [].
            obj.assertOpen();
            s = sentil_mex('ring_buffer_pop_back', obj.Handle);
        end

        function s = closest_to_time(obj, time)
            %CLOSEST_TO_TIME The held sample nearest a time, or [].
            obj.assertOpen();
            s = sentil_mex('ring_buffer_closest_to_time', obj.Handle, double(time));
        end

        function m = mean(obj)
            %MEAN The mean of the held values, or [] when empty.
            obj.assertOpen();
            m = sentil_mex('ring_buffer_mean', obj.Handle);
        end

        function v = variance(obj)
            obj.assertOpen();
            v = sentil_mex('ring_buffer_variance', obj.Handle);
        end

        function s = std_dev(obj)
            obj.assertOpen();
            s = sentil_mex('ring_buffer_std_dev', obj.Handle);
        end

        function m = min(obj)
            obj.assertOpen();
            m = sentil_mex('ring_buffer_min', obj.Handle);
        end

        function m = max(obj)
            obj.assertOpen();
            m = sentil_mex('ring_buffer_max', obj.Handle);
        end

        function recompute_statistics(obj)
            %RECOMPUTE_STATISTICS Recompute the running statistics from scratch.
            obj.assertOpen();
            sentil_mex('ring_buffer_recompute_statistics', obj.Handle);
        end

        function v = at_time(obj, time)
            %AT_TIME The value at an exact held time, or [].
            obj.assertOpen();
            v = sentil_mex('ring_buffer_at_time', obj.Handle, double(time));
        end

        function r = time_range(obj)
            %TIME_RANGE The [start, end] times spanned, or [] when empty.
            obj.assertOpen();
            r = sentil_mex('ring_buffer_time_range', obj.Handle);
        end

        function s = between(obj, startTime, endTime)
            %BETWEEN The samples with times in [start, end].
            obj.assertOpen();
            s = sentil_mex('ring_buffer_between', obj.Handle, double(startTime), double(endTime));
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this ring buffer has been closed');
            end
        end
    end
end